### ISSUE TITLE
Proposal: Add rules about (formatting of) namespace declarations

### ISSUE THREAD

**Issue by jrfnl** [WordPress/WordPress-Coding-Standards#1763]
## Is your feature request related to a problem?
PHP 5.3 introduced namespaces into PHP.

Refs:
* https://www.php.net/manual/en/migration53.new-features.php
* https://www.php.net/manual/en/language.namespaces.php

At this moment WPCS has no formatting rules for namespace declarations or the usage of namespaced constructs.

This issue, however, is purely about **namespace declarations**.

## Describe the solution you'd like
I would like to propose adding the following rules regarding namespace declarations to the WP Coding Standards handbook:
1. There should be exactly one space between the `namespace` keyword and the start of the namespace name in a namespace declaration.
2. Each "leaf" of a namespace name should be in `Camel_Caps`, i.e. each word should start with a capital and words should be separated by an underscore.
    Consecutive caps, like for acronyms, will be allowed.
    This is in line with other WP naming conventions.
3. There should be no whitespace or comments within the name part of the namespace declaration.
4. There should be exactly one blank line before a namespace declaration.
5. There should be at least one blank line after a namespace declaration.
    The "at least" bit prevent potential fixer conflicts with sniffs which check for a fixed amount of blank lines above function/class declarations.
6. There should be only one namespace declaration per file.
7. Scoped namespace declarations are not allowed.
8. Namespace declarations without name (= explicit global namespace) are not allowed.

### Valid code example
```php
<?php
/**
 * File docblock.
 */

namespace Prefix\Admin\Domain_URL\Sub_Domain\Event;

use ...
```

### Invalid code examples
```php
<?php
/**
 * File docblock.
 */
namespace Prefix \
	/* Here comes the real namespace */ Admin\DomainURL\SubDomain\Event;
use ...
```

```php
<?php
/**
 * File docblock.
 */
namespace Foo {
    // Code.
}

namespace {
    // Code.
}
```

If this proposal is accepted, we can then add sniff(s) to WPCS to check these rules automatically and - except for issues with comments - auto-fix them.

### Already covered
* The `namespace` keyword should be lowercase.
    This is already covered by the `Generic.PHP.LowerCaseKeyword.Found` sniff which is included in the `WordPress-Core` ruleset.

### Open question:
* Should we - at this time - define rules about the allowed compound depth of namespace names ? Or should we leave this open for now ?

### Explicitly not covered by this proposal:
* The name of a namespace should not start with a (superfluous) namespace separator.
    This would be a parse error in nearly all cases as PHP will in that case interpret the namespace name as the name of a global (namespaced) constant.
    Parse errors should be detected by linters and/or by the `Generic.PHP.Syntax` sniff.
* The order of various statements which can be at the start of a document.
    This will be covered in a separate proposal.
* Any relation between namespace declarations and autoloading / the path to the file.

### Comparison with PSR-12:
The only rule in the above proposal which is included in PSR-12 is rule 4 (one blank line before) and in a way rule 5 (at least one blank line after).
PSR-12 has no opinion on any of the other rules proposed here.

PSR-12 does have an opinion on the allowed compound namespace depth, but what with plugins and themes needing to use prefixing of namespaces, the PSR-12 rule for this may ultimately not be suitable for use in the context of WordPress.

### Comparison with PSR-1:
While PSR-1 (included in PSR-12) demands `StudlyCaps`/`PascalCase` for class names, it does not do so explicitly for namespace names.

PSR-1 does demand implementing PSR-0 or PSR-4 for autoloading, but as autoloading is explicitly not covered in this proposal, this is not something we currently need to take into account.

## Notes for the WPCS Implementation:
* Rule 1 can be covered by the upstream `Generic.WhiteSpace.LanguageConstructSpacing.IncorrectSingle` sniff. Also see issue #1616
* Rule 2 would need a dedicated WP sniff in the `NamingConventions` category.
* Rule 3 can be covered by an upstream sniff once it is pulled & merged. See: https://github.com/squizlabs/PHP_CodeSniffer/issues/2150
* Rule 4 can possibly be covered by the upstream PSR-12 sniff for the same, once it is created, pulled & merged.
* Rule 5 is already covered in `WordPress-Extra` by the upstream `PSR2.Namespaces.NamespaceDeclaration.BlankLineAfter` sniff. This sniff could be moved to `WordPress-Core`.
* Rule 6, 7 and 8 can be covered by upstream sniffs once pulled & merged. See: https://github.com/squizlabs/PHP_CodeSniffer/issues/2324

/cc @pento

**Comment by pento** [comment]
My name is Gary, and I approve of this proposal.

**Comment by GaryJones** [comment]
> words should be separated by an underscore.

Please make the check for use of underscores versus PascalCase be optionally disable-able - folks may want to go against that for their directory name:

```
namespace PHPUnit\Framework\MockObject;
```

Or project name:

```
namespace Automattic\WPCOMLegacyRedirector;
```

Or vendor (brand) name:

```php
namespace WordPress\Foo;
```
```php
namespace StudioPress\Genesis;
```

> Should we - at this time - define rules about the allowed compound depth of namespace names ? Or should we leave this open for now ?

Leave open for now, at least for the wider standards. If WP Core wants to limit to a certain depth, then I see now harm in building it in so long as the default equates to no check happening (depth 0 = no check).

Otherwise:

> My name is Gary, and I approve of this proposal.

**Comment by jrfnl** [comment]
> Please make the check for use of underscores versus PascalCase be optionally disable-able

Per the above proposal, that rule (2) would be covered in a separate `NamingConventions` sniff, which then could be quite easily turned off, both from within a custom ruleset, as well as from the command-line.

**Comment by kwhat** [comment]
Rule 7 is going to be problematic for unit testing.  I have outlined my argument in [squizlabs/PHP_CodeSniffer#2324](https://github.com/squizlabs/PHP_CodeSniffer/issues/2324#issuecomment-532817110).

**Comment by jrfnl** [comment]
@kwhat I am probably missing something, but I honestly don't see your problem.

You can just have the mock functions in a separate file with its own (unscoped) namespace and either include that file from your test bootstrap or via an autoloader and AFAICS things should work just fine. What am I missing ?

On a related note: there's nothing stopping you from excluding the sniff (once added to WPCS) for your projects and/or selectively excluding it for the unit tests in your project.

The fact that you _choose_ to do something which is counter to WPCS, is no reason not to add the rule to WPCS.

**Comment by azaozz** [comment]
Just a note that namespace declarations are not part of the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/) (yet), see https://make.wordpress.org/core/2020/03/20/updating-the-coding-standards-for-modern-php/. 

This enhancement does not belong in WPCS unless namespace declarations are added to the coding standards.

**Comment by odil-io** [comment]
I found this while researching and thought it might help someone:

tl:dr; WPCS uses PSR-2 as of writing.

According to the [WordPress Coding Standards](https://github.com/WordPress/WordPress-Coding-Standards), the package [squizlabs/PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer) is used to run PHPCS with extra WordPress rules.
By default, `squizlabs/PHP_CodeSniffer` Sniffs code against the PSR-2 standard. See  [CodeSniffer.conf.dist](https://github.com/squizlabs/PHP_CodeSniffer/blob/master/CodeSniffer.conf.dist).

Within the WordPress-Core subpackage we can see [here ](https://github.com/WordPress/WordPress-Coding-Standards/blob/develop/WordPress-Core/ruleset.xml) that it used the [PSR-2 Style Guide](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md). But:

> Deprecated - As of 2019-08-10 PSR-2 has been marked as deprecated. [PSR-12](https://www.php-fig.org/psr/psr-12/) is now recommended as an alternative.

I guess we still have to wait a bit longer for WPCS to include their rules + those from PSR-12.

**Comment by jrfnl** [comment]
@odil-io Sorry, your research is misleading you. Also, please don't post these kind of comments in a legitimate discussion as they are unrelated to what is being discussed here.

> tl:dr; WPCS uses PSR-2 as of writing.

No, WPCS has its own set of rules which are described in the [Handbook](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/).

The PSR2 standard from PHPCS is **not** used, though select sniffs from that standard _are_ used.,

> By default, `squizlabs/PHP_CodeSniffer` Sniffs code against the PSR-2 standard. See [CodeSniffer.conf.dist](https://github.com/squizlabs/PHP_CodeSniffer/blob/master/CodeSniffer.conf.dist).

No, by default - if no `--standard` or `[.]phpcs.xml[.dist]` file is provided and the native default is not overruled via a `codeSniffer.conf` files - PHPCS scans against the [PEAR standard](https://github.com/squizlabs/PHP_CodeSniffer/wiki/Configuration-Options#setting-the-default-coding-standard).
PHPCS 4.x will change this to PSR12, but this change will have no effect on WPCS users as WPCS is a standard in its own right and provides its own ruleset with rules.

> Within the WordPress-Core subpackage we can see [here ](https://github.com/WordPress/WordPress-Coding-Standards/blob/develop/WordPress-Core/ruleset.xml) that it used the [PSR-2 Style Guide](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md).

Again, No. The ruleset you point to does not include `<rule ref="PSR2"/>`, so is not based on PSR2.

**Comment by odil-io** [comment]
@jrfnl Sorry, I wasn't paying enough attention. I made wrong assumptions, specifically in this:

> WPCS is a standard in its own right and provides its own ruleset with rules.

By that, anything in any PSR standard _does not apply_ for WPCS. 

Having said that I hereby end my participation to this discussion as I believe I wasn't supposed to comment on this issue, at all.

**Comment by jrfnl** [comment]
@odil-io You're welcome to participate in any discussion in this repo, but please restrict comments in open issues to the topic being discussed. If you can't find an open issue which covers the topic, please open a new issue for whatever you want to discuss (as long as it is related to WPCS).
